canonical-kubernetes/steps/02_get-kubectl/after-deploy

```diff
@@ -1,29 +1,53 @@
 #!/bin/bash
 
-set -eux
+set -e
 
-. $CONJURE_UP_SPELLSDIR/sdk/common.sh
+. "$CONJURE_UP_SPELLSDIR/sdk/common.sh"
 
-mkdir -p $HOME/.kube
-mkdir -p $HOME/bin
-
-juju scp -m $JUJU_CONTROLLER:$JUJU_MODEL kubernetes-master/0:config ~/.kube/config.$JUJU_MODEL
+mkdir -p "$HOME/.kube"
 
 if [[ $(uname -s) = "Darwin" ]]; then
     KUBE_DEST="/usr/local/bin"
     KUBE_REPO=https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/darwin/amd64
 
-    curl -sLO $KUBE_REPO/kubectl
+    curl -sLO "$KUBE_REPO/kubectl"
     chmod +x kubectl
-    mv kubectl $KUBE_DEST/kubectl
+    mv kubectl "$KUBE_DEST/kubectl"
 else
     KUBE_DEST="/snap/bin"
 
     sudo snap install kubectl --classic || true
 fi
 
-echo "KUBECONFIG=$HOME/.kube/config.$JUJU_MODEL $KUBE_DEST/kubectl \$@" > $HOME/bin/kubectl
-chmod +x $HOME/bin/kubectl
+# copy down config file from cluster
+juju scp -m "$JUJU_CONTROLLER:$JUJU_MODEL" kubernetes-master/0:config "$HOME/.kube/config.conjure-up.$JUJU_MODEL"
+
+# change cluster, context, and name of user to model name
+"$(dirname "$0")/update_kube_config.py" "$HOME/.kube/config.conjure-up.$JUJU_MODEL" "$JUJU_MODEL"
+
+# merge with any existing config and dump to config.conjure-up.new
+KUBECONFIG="$HOME/.kube/config.conjure-up.$JUJU_MODEL:$HOME/.kube/config" "$KUBE_DEST/kubectl" config view --raw=true > "$HOME/.kube/config.conjure-up.new"
+
+# manage backup if existing config
+if [[ -e "$HOME/.kube/config" ]]; then
+    # make backup copy of old config
+    cp "$HOME/.kube/config" "$HOME/.kube/config.conjure-up.backup.$(date -Iseconds)"
+    # purge backups, keeping most recent 10
+    for old_backup in $(find "$HOME/.kube" -name 'config.conjure-up.backup.*' | sort | head -n-10); do
+        rm "$old_backup" || true  # ignore errors
+    done
+fi
+
+# replace any existing config with merged, and make default
+mv "$HOME/.kube/config.conjure-up.new" "$HOME/.kube/config"
```

Comment:
Could we not overwrite and instead merge the config of the new config with the configs of any preexisting clusters?  Something similar to https://github.com/ktsakalozos/blog-post-federation/blob/master/scripts/merge-configs.py would help. This way the user will not lose access to any clusters he may already have and he will be able to switch between clusters via `kubectl config set current-context <cluster_name>`. Also, having multiple cluster in the config is needed in the case we want to support federation.

Reply:
You missed [line 29](https://github.com/conjure-up/spells/pull/173/files#diff-b70646b46247af76e5fe8848cca335d5R29).  It does merge the config and use a context to identify the cluster, I just used `kubectl` to perform the merge rather than modifying the file directly.  I didn't realize you had already written a merge script, or I would have used it, though I do feel a bit more comfortable with using `kubectl` to write the "official" file.